src/main/java/com/neuronrobotics/bowlerbuilder/controller/WebBrowserController.java

```diff
@@ -68,8 +174,121 @@ private void onNavigate(ActionEvent actionEvent) {
     loadPage(url);
   }
 
-  public void loadPage(String url) {
+  @FXML
+  private void onRun(ActionEvent actionEvent) {
+    Thread thread = LoggerUtilities.newLoggingThread(logger, () -> {
+      try {
+        final File currentFile = ScriptingEngine.fileFromGit(currentGist,
+            fileBox.getSelectionModel().getSelectedItem());
+        final String name = currentFile.getName();
+        final Object obj = ScriptingEngine.inlineScriptRun(currentFile, null,
+            ScriptingEngine.getShellType(name));
+        parseResult(obj);
+      } catch (Exception e) {
+        logger.warning("Could not parse and run script.\n"
+            + Throwables.getStackTraceAsString(e));
+      }
+    });
+    thread.setDaemon(true);
+    thread.start();
+  }
+
+  /**
+   * Parse the result of a script. CSG objects get added to a CAD engine.
+   *
+   * @param result script result
+   */
+  private void parseResult(@Nullable final Object result) {
+    if (result instanceof Iterable) {
+      final Iterable<?> iterable = (Iterable) result;
+      final Object firstElement = iterable.iterator().next();
+
+      if (firstElement instanceof CSG) {
+        parseCSG((Iterable<CSG>) iterable);
+      }
+    }
+  }
+
+  /**
+   * Add CSGs to a new {@link AceCadEditorTab}.
+   *
+   * @param csgIterable CSGs to add
+   */
+  private void parseCSG(@Nonnull Iterable<CSG> csgIterable) {
+    Platform.runLater(() -> {
+      final String selection = fileBox.getSelectionModel().getSelectedItem();
+      try {
+        AceCadEditorTab tab = new AceCadEditorTab(selection);
+        tab.getController().getCADViewerController().addAllCSGs(csgIterable);
+        final GHGist gist = ScriptingEngine.getGithub()
+            .getGist(ScriptingEngine.urlToGist(currentGist));
+        tab.getController().loadGist(gist, gist.getFile(selection));
+        parentController.addTab(tab);
+      } catch (IOException e) {
+        e.printStackTrace();
+      }
+    });
+  }
+
+  @FXML
+  private void onModify(ActionEvent actionEvent) {
+    final String selection = fileBox.getSelectionModel().getSelectedItem();
+
+    try {
+      final File currentFile = ScriptingEngine.fileFromGit(currentGist, selection);
+      //TODO: Doesn't properly detect owner
+      final Boolean isOwner = ScriptingEngine.checkOwner(currentFile);
```

Comment:
@madhephaestus Can you comment on how to properly detect the owner of a file? Doing this doesn't seem to work.

Reply:
can you tell me if this is it failing true or false? It is checking the logged in user if it has push access by pushing the unchanged repository. If it has permission it will succeed. If anything fails, or the repo had been changed somehow and not committed it might false-fail.

Reply:
It never returns `true`.

Reply:
Ok, the debugger tells me `currentGist` is `https://gist.github.com/d4312a0787456ec27a2a.git` and `currentFile` is `helloWorld.groovy` when I load the page `http://commonwealthrobotics.com/BowlerStudio/Welcome-To-BowlerStudio/`

Reply:
well, i own that gist. it shouldn't return true for you.

Reply:
So the file must be wrong then. After I hit the "Make a copy" button, it still returns false. I copy the gist by:
`final GHGist gist = ScriptingEngine.fork(ScriptingEngine.urlToGist(currentGist));`
(then it gets opened in an editor)

After I copy it like that, will `ScriptingEngine.fileFromGit(currentGist, file)` still use your gist or my fork of it? Maybe I'm not correctly updating `currentGist` and `file` after I make the fork.

Reply:
this sounds like a stale variable state maybe?